### ISSUE TITLE
Debugger caused the quitting of some other plugins

### ISSUE THREAD

**Issue by YvesZHI** [fatih/vim-go#3020]
### What did you do?
I installed Vim-go and other plugins, such as Nerdtree for my vim.
I entered my vim with all of plugins executed as expected, such as Nerdtree.
Then I typed `:GoDebugStart .` to try to debug with Vim-go.

Before typing `:GoDebugStart .`, one window for the file `main.go` and one window for the plugin Nerdtree are opened. The cursor is in the `main.go`.
![image](https://user-images.githubusercontent.com/27804279/94384269-b9cd3680-0174-11eb-9663-e0a7b4b53319.png)

After executing `:GoDebugStart .`, the window of Nerdtree disappeared and the cursor is still in the `main.go`.
![image](https://user-images.githubusercontent.com/27804279/94384474-39f39c00-0175-11eb-9402-eb668220328c.png)

Then when I execute `:GoDebugStop`, the window of Nerdtree won't come back:
![image](https://user-images.githubusercontent.com/27804279/94384584-8048fb00-0175-11eb-8803-c4a2b8a45ac5.png)

### What did you expect to happen?
When I type `:GoDebugStart .`, the window of Nerdtree shouldn't be closed. Or, when I type `:GoDebugStop` to quit the debugger of Vim-go, I was expecting that the window of Nerdtree could come back.

### What happened instead?
When I type `:GoDebugStart .`, the window of Nerdtree is closed. When I type `:GoDebugStop` to quit the debugger of Vim-go, the window of Nerdtree won't come back.

#### vim-go version:
commit ef23d90b84231759a992adf79b0dcd72844c1f84
####  `vimrc` you used to reproduce:

<details><summary>vimrc</summary><br><pre>
    
    set rtp+=~/.vim/bundle/Vundle.vim
    call vundle#begin()

    " let Vundle manage Vundle, required
    Plugin 'VundleVim/Vundle.vim'

    " options
    "Plugin 'Valloric/YouCompleteMe'
    Plugin 'scrooloose/nerdtree'
    Plugin 'fatih/vim-go'
    Plugin 'ctrlpvim/ctrlp.vim'
    Plugin 'scrooloose/nerdcommenter'
    Plugin 'majutsushi/tagbar'
    Plugin 'vim-airline/vim-airline'
    Plugin 'vim-airline/vim-airline-themes'
    Plugin 'rstacruz/sparkup', {'rtp': 'vim/'}
    Plugin 'Yggdroot/indentLine'
    Plugin 'jiangmiao/auto-pairs'
    Plugin 'tpope/vim-repeat'
    Plugin 'YvesZHI/vim-code-dark'
    Plugin 'YvesZHI/killersheep'
    call vundle#end()

</pre></details>

#### Vim version (first three lines from `:version`):

VIM - Vi IMproved 8.2
#### Go version (`go version`):

go version go1.14.6 linux/amd64
#### Go environment
<details><summary><code>go env</code> Output:</summary><br><pre>

GO111MODULE="auto"
GOARCH="amd64"
GOBIN=""
GOCACHE="/home/zyh/.cache/go-build"
GOENV="/home/zyh/.config/go/env"
GOEXE=""
GOFLAGS=""
GOHOSTARCH="amd64"
GOHOSTOS="linux"
GOINSECURE=""
GONOPROXY=""
GONOSUMDB=""
GOOS="linux"
GOPATH="/home/zyh/go"
GOPRIVATE=""
GOPROXY="https://goproxy.cn,direct"
GOROOT="/usr/local/go"
GOSUMDB="sum.golang.org"
GOTMPDIR=""
GOTOOLDIR="/usr/local/go/pkg/tool/linux_amd64"
GCCGO="gccgo"
AR="ar"
CC="gcc"
CXX="g++"
CGO_ENABLED="1"
GOMOD=""
CGO_CFLAGS="-g -O2"
CGO_CPPFLAGS=""
CGO_CXXFLAGS="-g -O2"
CGO_FFLAGS="-g -O2"
CGO_LDFLAGS="-g -O2"
PKG_CONFIG="pkg-config"
GOGCCFLAGS="-fPIC -m64 -pthread -fmessage-length=0 -fdebug-prefix-map=/tmp/go-build308368298=/tmp/go-build -gno-record-gcc-switches"

</pre></details>

#### gopls version
<details><summary><code>gopls version</code> Output:</summary><br><pre>

golang.org/x/tools/gopls 0.4.3
</pre></details>

**Comment by bhcleek** [comment]
Can you give me explicit instructions to duplicate this? I cannot duplicate the general case that you describe, but I think I might see a potential cause for this in the code; explicit steps to duplicate would help. Specifically, I'd need to know:
* what windows to have open when `:GoDebugStart` is run and which window the cursor should be in
* what windows to have open when `:GoDebugStop is run and which window the cursor should be in

**Comment by YvesZHI** [comment]
@bhcleek   I've re-edited the issue. Please check it out.

**Comment by bhcleek** [comment]
It appears that your screenshots show that the NerdTree window is closed on `:GoDebugStart`, not on `:GoDebugStop` which is contrary to your description. Can you you clarify which behavior you're seeing?

**Comment by YvesZHI** [comment]
@bhcleek   Well, you are right. but if `GoDebugStart` makes the Nerdtree close, it is OK for me. What I expected is that if I execute `GoDebugStop`, the Nerdtree should come back.

**Comment by YvesZHI** [comment]
@bhcleek  I just re-edited the issue again.
So I'm expecting that `:GoDebugStart .` won't close the window of Nerdtree, or `:GoDebugStop` can bring the window of Nerdtree back to me.

**Comment by bhcleek** [comment]
Vim-go makes the active buffer then only window when it starts editing and then opens a windows to facilitate debugging. I don't have any current plans to change that, but I'll consider making a change before the next release.

**Comment by bhcleek** [comment]
We might be able to use `:mksession` to save the current window and buffer layout and then use `:source` at the end of debugging to restore the session 🤔 

@YvesZHI you do not have to wait for this to be added to vim-go: you can run `:mksession` now before starting debugging and then use `:source` to reload the session file.

**Comment by bhcleek** [comment]
I don't see a good way to change the current behavior. Here are some options, all with drawbacks:

#### use a new tab for debugging

When the debugger starts up, open a new tab with the current buffer loaded and the additional debugging windows loaded into to the new tab.  When debugging stops close the tab and all tabs in it.

The disadvantage here is that the user could be in one buffer at the last bit of debugging, and then be suddenly thrown back into their original buffer when debugging stops.

#### use sessions

When the debugger starts, using `:mksession` to save a session file and restore that session when debugging stops.

The disadvantage here is that the user could be in one buffer at the last bit of debugging, and then be suddenly thrown back into their original buffer when debugging stops.

#### split the current window when debugging starts

Don't use `:only` when the debugger starts.

This would be an awful experience for your use case; there's just not enough screen real estate in a console to keep all of the existing windows and then perform multiple splits in the current window when debugging starts.

#### conclusion

I am very tempted to close this as a won't fix, because the current behavior seems like the best tradeoff given all the options, but I'm open to other ideas.

**Comment by YvesZHI** [comment]
@bhcleek  Thanks for what you've done. I don't quite understand how Vim 8 works but Vim 8 supports its own debugger: `:h Termdebug`, it seems that GDB support-ONLY. When I execute `:Termdebug` while coding C/C++, it works as expected, meaning that when I quit the debugger, it won't close other plugin windows, such as Nerdtree.

I don't know if it is helpful for this issue.

I will close this issue in few days.

**Comment by bhcleek** [comment]
There is another way to address your concern: you can use `g:go_debug_windows` to cause vim-go to not open the additional windows. We'd need to add some simple feature option to avoid running `:only` when `g:go_debug_windows` is empty. WDYT?

**Comment by bhcleek** [comment]
@YvesZHI do you want to give #3068 a try with `let g:go_debug_windows= {}` in your vimrc?

**Comment by bhcleek** [comment]
@YvesZHI bump

**Comment by patapra** [comment]
> #### split the current window when debugging starts
> Don't use `:only` when the debugger starts.
> 
> This would be an awful experience for your use case; there's just not enough screen real estate in a console to keep all of the existing windows and then perform multiple splits in the current window when debugging starts.

@bhcleek Sorry for commenting on a closed issue (I can create a new one if you'd like), but I am running into this issue now and would like to propose a different solution based on your comment above:

Empower the dev to choose how to best use their screen real estate by adding an option to not use `:only` when debugging. The default for the option would retain the current behavior. It could be an optional key in the `g:go_debug_windows` dict. If not present, then assume `true`:
```
let g:go_debug_windows = {
    \ 'vars':       'rightbelow 30vnew',
    \ 'stack':      'rightbelow 20new',
    \ 'goroutines': 'botright 10new',
    \ 'out':        'botright 5new',
    \ 'only':        false,
\ }
```
WDYT?

**Comment by bhcleek** [comment]
Sure. that would be easy enough to do. Feel free to submit a PR.

**Comment by patapra** [comment]
Created #3125
